File: server/src/main/java/org/dreamexposure/discal/server/api/endpoints/v2/event/list/ListEventRangeEndpoint.java
package org.dreamexposure.discal.server.api.endpoints.v2.event.list;

import org.dreamexposure.discal.core.database.DatabaseManager;
import org.dreamexposure.discal.core.logger.LogFeed;
import org.dreamexposure.discal.core.logger.object.LogObject;
import org.dreamexposure.discal.core.object.GuildSettings;
import org.dreamexposure.discal.core.object.web.AuthenticationState;
import org.dreamexposure.discal.core.utils.GlobalConst;
import org.dreamexposure.discal.core.utils.JsonUtil;
import org.dreamexposure.discal.core.utils.JsonUtils;
import org.dreamexposure.discal.core.wrapper.google.EventWrapper;
import org.dreamexposure.discal.server.utils.Authentication;
import org.json.JSONException;
import org.json.JSONObject;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import discord4j.common.util.Snowflake;
import reactor.core.publisher.Flux;

@RestController
@RequestMapping("/v2/events/list")
public class ListEventRangeEndpoint {
    @PostMapping(value = "/range", produces = "application/json")
    public String getEventsForRange(final HttpServletRequest request, final HttpServletResponse response, @RequestBody final String rBody) {
        //Authenticate...
        final AuthenticationState authState = Authentication.authenticate(request);
        if (!authState.getSuccess()) {
            response.setStatus(authState.getStatus());
            response.setContentType("application/json");
            return JsonUtil.INSTANCE.encodeToString(AuthenticationState.class, authState);
        }

        //Okay, now handle actual request.
        try {
            final JSONObject requestBody = new JSONObject(rBody);

            final Snowflake guildId = Snowflake.of(requestBody.getString("guild_id"));
            final int calNumber = requestBody.getInt("calendar_number");
            final long startEpoch = requestBody.getLong("epoch_start");
            final long endEpoch = requestBody.getLong("epoch_end");
            final GuildSettings settings = DatabaseManager.getSettings(guildId).block();

            //okay, lets actually get the range's events.
            final List<JSONObject> events = DatabaseManager.getCalendar(settings.getGuildID(), calNumber)
                .flatMap(calData -> EventWrapper.getEvents(calData, startEpoch, endEpoch))
                .flatMapMany(Flux::fromIterable)
                .map(e -> JsonUtils.convertEventToJson(e, settings))
                .collectList()
                .block();

            final JSONObject body = new JSONObject();
            body.put("events", events);

            response.setContentType("application/json");
            response.setStatus(GlobalConst.STATUS_SUCCESS);
            return body.toString();
        } catch (final JSONException e) {
            e.printStackTrace();

            response.setContentType("application/json");
            response.setStatus(GlobalConst.STATUS_BAD_REQUEST);
            return JsonUtils.getJsonResponseMessage("Bad Request");
        } catch (final Exception e) {
            LogFeed.log(LogObject
                .forException("[API-v2]", "get events for range err", e, this.getClass()));

            response.setContentType("application/json");
            response.setStatus(GlobalConst.STATUS_INTERNAL_ERROR);
            return JsonUtils.getJsonResponseMessage("Internal Server Error");
        }
    }
}

File: server/src/main/kotlin/org/dreamexposure/discal/server/endpoints/v2/event/list/ListEventRangeEndpoint.kt
package org.dreamexposure.discal.server.endpoints.v2.event.list

import com.fasterxml.jackson.databind.ObjectMapper
import discord4j.common.util.Snowflake
import kotlinx.coroutines.reactor.mono
import kotlinx.serialization.encodeToString
import org.dreamexposure.discal.core.annotations.SecurityRequirement
import org.dreamexposure.discal.core.business.CalendarService
import org.dreamexposure.discal.core.logger.LOGGER
import org.dreamexposure.discal.core.`object`.new.model.discal.v2.EventV2Model
import org.dreamexposure.discal.core.utils.GlobalVal
import org.dreamexposure.discal.server.utils.Authentication
import org.dreamexposure.discal.server.utils.responseMessage
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import org.springframework.http.server.reactive.ServerHttpResponse
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.server.ServerWebExchange
import reactor.core.publisher.Mono
import java.time.Instant

@RestController
@RequestMapping("/v2/events/list")
class ListEventRangeEndpoint(
    private val authentication: Authentication,
    private val calendarService: CalendarService,
    private val objectMapper: ObjectMapper,
) {
    @PostMapping("/range", produces = ["application/json"])
    @SecurityRequirement(disableSecurity = true, scopes = [])
    fun listByRange(swe: ServerWebExchange, response: ServerHttpResponse, @RequestBody rBody: String): Mono<String> {
        return authentication.authenticate(swe).flatMap { authState ->
            if (!authState.success) {
                response.rawStatusCode = authState.status
                return@flatMap Mono.just(GlobalVal.JSON_FORMAT.encodeToString(authState))
            }

            //Handle request
            val body = JSONObject(rBody)
            val guildId = Snowflake.of(body.getString("guild_id"))
            val calendarNumber = body.getInt("calendar_number")
            val start = Instant.ofEpochMilli(body.getLong("epoch_start"))
            val end = Instant.ofEpochMilli(body.getLong("epoch_end"))

            mono { calendarService.getCalendar(guildId, calendarNumber) }.flatMap { calendar ->
                mono { calendarService.getEventsInTimeRange(guildId, calendarNumber, start, end) }.map { events ->
                    events.map { objectMapper.writeValueAsString(EventV2Model(it, calendar)) }
                }
            }.map(::JSONArray)
                .map { JSONObject().put("events", it).put("message", "Success").toString() }
                .doOnNext { response.rawStatusCode = GlobalVal.STATUS_SUCCESS }
                .switchIfEmpty(responseMessage("Calendar not found")
                    .doOnNext { response.rawStatusCode = GlobalVal.STATUS_NOT_FOUND }
                )
        }.onErrorResume(JSONException::class.java) {
            LOGGER.trace("[API-v2] JSON error. Bad request?", it)

            response.rawStatusCode = GlobalVal.STATUS_BAD_REQUEST
            return@onErrorResume responseMessage("Bad Request")
        }.onErrorResume {
            LOGGER.error(GlobalVal.DEFAULT, "[API-v2] list events by range error", it)

            response.rawStatusCode = GlobalVal.STATUS_INTERNAL_ERROR
            return@onErrorResume responseMessage("Internal Server Error")
        }
    }
}